### ISSUE TITLE
Feature/Package Module Proposal

### ISSUE THREAD

**Issue by behindthepage** [backdrop/backdrop-issues#1238]
Backdrop needs a system to enable modules to easily and safely add views and layouts to a sites config folder. Then modules can add "Features" with a combination of views and layouts config files.

This can be done now by having the views and layouts in the modules /config folder which are automatically added to the sites config folder. An example of a WP blog "features" modules that will do this is here https://github.com/behindthepage/blog
But the danger with this approach is that these files may over write files with the same name or create views and layouts with duplicate paths and the views and layouts are not removed when the module is disabled and uninstalled.

I envisage a system where a module can declare it's "addon" views and layouts and when enabled Backdrop checks that there are no naming or path issues and when disabled Backdrop disabled the views and layouts and deletes them when uninstalled maybe asking the user to confirm the deletion.

To make this work I suggest adding a hook_register_addons() function to the module's install file and having an extra folder in the module folder for the addon views and layouts.

So to modify my blog example 
The views and layout files would be in blog/addons

blog/addons/layout.layout.news.json 
blog/addons/views.view.archives.json
blog/addons/views.view.news.json
blog/addons/views.view.tags.json

and the function in the install file would be

``` php
function blog_register_addons() {

  $addons['blog'] = array(
    'description' => 'Provides a blog and archive pages.',
    'config_files' => array(
       'views.view.blog_news' => array(
        'description' => 'Provides a blog view.',  
      ),
      'views.view.blog_archives' => array(
        'description' => 'Provides an archive page and archive navigation block.',
      ),
      'views.view.blog_tags' => array(
        'description' => 'Provides a tags block.',
      ),
      'layout.layout.blog_news' => array(
        'description' => 'Provides a loyout for the blog page at /news.',
      ),
    ),
  );

  return $addons;
}   
```

When the module is enables backdrop would check that there is no naming issues, warn of any path issues and copy these files into the sites config folder.

When the module is disabled or uninstalled a information pane would appear explaining what the views and layouts did and asking for confirmation that thay can be deleted. It could also explain to get them back you can just enable the module again.

I have used the name "addons" but it could be 
blog_register_feature() 
blog_register_package() 
or something else.

Related discussions:
https://github.com/backdrop/backdrop-issues/issues/1045
https://github.com/backdrop/backdrop-issues/issues/881

What does everyone think?

Regards
Geoff

**Comment by sutibun** [comment]
The monkey likes.

> But the danger with this approach is that these files may over write files with the same name or create views and layouts with duplicate paths and the views and layouts are not removed when the module is disabled and uninstalled.

What if to prevent naming conflicts, the views and layout name defaults to "MODULE_NAME_CAPITALIZED" + "_module:" + "DESIRED_NAME"

So in the case of your blog module, you'd have "BLOG_MODULE:blog" (assuming you'd want a view called blog.)

In this way, module-created views or layouts stand out. 

Or, maybe another alternative is separating module created views and layouts from the ones created by the user.

**Comment by behindthepage** [comment]
Hi Monkey ;-)

The config file names have strict naming conventions already. The first is the module that owns them and all views are owned by views module and layouts by layouts module. the next is what it is e.g. view or layout then comes the name of the view or layout so we don't have any wiggle room there.

Regards
Geoff

**Comment by sutibun** [comment]
> The config file names have strict naming conventions already.

😳 \*slowly backs away & exits stage right\*

**Comment by quicksketch** [comment]
> When the module is enables backdrop would check that there is no naming issues, warn of any path issues and copy these files into the sites config folder.

I think this would be a good idea in any case. We shouldn't allow modules to be enabled at all if they have conflicting config file names with any existing config.

> When the module is disabled or uninstalled a information pane would appear explaining what the views and layouts did and asking for confirmation that thay can be deleted. It could also explain to get them back you can just enable the module again.

All config files owned by a module are already uninstalled, but I see how this is different in that config files that aren't necessarily owned by a module would also be removed. For example "my_module.module" might provide "my_module.settings.json", which would be automatically installed. But "views.view.my_module_view.json" would not. because it's namespace-wise owned by Views, not my_module.

The thing I'm a little puzzled by in this proposal is that you already know what default config files a module ships with, don't you? They're the files that are in the module's `config` directory. Although the hook could provide some human-readable names, it seems like you'd already have the information needed to uninstall related config files.

**Comment by behindthepage** [comment]
Hi Nate,

The hook also provides a description that can be used to warn the user what will be lost when a module is disabled. I also thought of a senario of someone wanting to keep one of the views or layout and being given the choice when disabling a module but that might be over thinking it. They could just clone the view they wanted to keep. We know these things are happening but for a newb things sometime disappear without seemly any reason. Although a  good readme file or backdrop_set_message may all that may be needed 

If views and layouts are checked for naming and path issues before being copied that would solve the main problem here. 

It is easy enough to delete any config files in hook uninstall

```php
function blog_uninstall() {
  // remove added config files
  config('layout.layout.blog_news')->delete();
  config('views.view.blog_archives')->delete();
  config('views.view.blog_news')->delete();
  config('views.view.blog_tags')->delete();
  cache_flush('cache');
}
```
The rest may just be me getting too fancy.

Is there anywhere in the documentation that explains how a module can add a view or layout?
Maybe an example module would be helpful. I would be happy to do this once we have the namespace issue solved. My blog module with a bit of polish may be suitable.

Regards
Geoff

**Comment by docwilmot** [comment]
@behindthepage nice to see you havent given up on this. Its something that I dream that we could standardize eventually. Thoughts re this:

- It would be good if a module which provides this sort of functionality could declare itself as a "packager" in some way. A blog as in your example may be a simple thing, but if this becomes standard, developers may start wanting to package larger concepts with multiple module dependencies, config files, views, content types, etc. We'd need to be able to let the end user know that this is going to happen and that this isn't just a simple "module." This may jive well with [my ongoing campaign](https://github.com/backdrop-ops/backdropcms.org/issues/100) to add "tags[]" to module `.info` files.

- We have the intention still of including Project module in core. If we got it to be able to recognize the concept of a "package" it would be great UX to just search for a "School site" package and just download and install and let the package and PM do all the work.

- Re the new proposed hook, couldnt we just add this dupe-checking functionality to `hook_config_info()`?

In conclusion I still like this idea.

**Comment by quicksketch** [comment]
> We have the intention still of including Project module in core

I think you meant, Project *Browser* module. :smile:

> Is there anywhere in the documentation that explains how a module can add a view or layout?

Hm, I don't think so, but I know the API module just ships with a couple of views: https://github.com/backdrop-contrib/api/tree/1.x-1.x/config

However it doesn't remove them on uninstall, but perhaps it should.

**Comment by klonos** [comment]
> I have used the name "addons" but it could be
> blog_register_feature()
> blog_register_package()
> or something else.

I like "features" or "extras" perhaps.

**Comment by docwilmot** [comment]
Heres another discussion about similar issues: https://github.com/backdrop/backdrop-issues/issues/99.
I think this has been discussed to death already, can we start doing it now?
What do we need to do or know to start?

**Comment by quicksketch** [comment]
>  I think this has been discussed to death already, can we start doing it now?

I'm not actually what it is that needs doing. I think @behindthepage's suggestions here might add a bit of formality to the idea of features, but I don't like the similarity in purpose between `hook_config_info()` and `hook_register_addons()` (or whatever the new hook would be).

I think we've already got the tools we need between config and install/uninstall hooks to do the essentials of a module that bundles together various bits of configuration to assemble a "feature". Increasingly, that's what even our own core modules are becoming. Consider Node and User module's use of Views for administration for example. Of course those you can't disable, but a module like https://backdropcms.org/project/api includes 5 views and a content type.

If all we're talking about is making a module that includes a bunch of default configurations and manages adding/removing them when the module is enabled/disabled, we already have the tools needed to do such a thing, but each module may end up doing it differently (which might be okay and even necessary, depending on what the module provides).

**Comment by behindthepage** [comment]
This thread has sort of gone off topic and is discussing a larger issue than I first raised. A good issue, but a very large one which would be better discussed in one of the already started threads . Lets solve one issue at a time.

I concede that we probably don't need a new hook to achieve solutions to the issues I am raising so I am down to 3 suggestions. 

1.  I think we have agreed that:

>We shouldn't allow modules to be enabled at all if they have conflicting config file names with any existing config.

That is something that we can fix now. Yay!

It would be nice if instead of a message like "This module can't be enabled because the config file xyz already exists. Have a nice day" there was some way to solve this conflict so the user wouldn't have to worry about solving this problem. However this may be a rare occurrence and the message should just say start an issue for this module.

2. I think paths should be checked when views or layouts are being installed. I think it is more likely to duplicate a path when installing a view or a layout than having conflicting config file names and this too can be solved with a bit more code during the copying of the config files. Maybe we can change the new path to duplicate_0 and set a message during install.

3. It would be nice to have some documentation and an example module showing how a module can add views and layouts which I would be happy to do.

I think these three things would improve the whole "module adding a feature" situation.

Regards
Geoff

**Comment by quicksketch** [comment]
> It would be nice if instead of a message like "This module can't be enabled because the config file xyz already exists. Have a nice day"

https://github.com/backdrop/backdrop-issues/issues/1242

> I think paths should be checked when views or layouts are being installed.

Let's make an issue for this one in particular. I think we may be able to do a check, but doing a validation before installation may be difficult because each type of thing makes a menu path differently. It's only when they're all consolidated into `hook_menu()` that you have the full list. And then each module could have duplicates within it as well (two views at the same path, two layouts at the same path, etc.)

> It would be nice to have some documentation and an example module showing how a module can add views and layouts which I would be happy to do.

That's great. I would suggest putting it in https://github.com/backdrop-contrib/examples project, which is also parsed and displayed on http://api.backdropcms.org. A stand-alone documentation page might be a good idea as well.

**Comment by behindthepage** [comment]
Thanks Nate. I will work on this when I get some time.

Regards
Geoff

**Comment by docwilmot** [comment]
Great summary @behindthepage.

**Comment by klonos** [comment]
>>I think paths should be checked when views or layouts are being installed.

> Let's make an issue for this one in particular.

Done.

**Comment by behindthepage** [comment]
Thanks @klonos

**Comment by klonos** [comment]
No worries @behindthepage - just doing my part in the queue.

Hey, I just saw that you are NSW-based! I'm in VIC myself :wink: ...been here for about a year and a half, but initially landed in Sydney and stayed there for 3 months before moving to Melbourne.
